source/request.test.js

```diff
@@ -0,0 +1,212 @@
+const Request = require('./request');
+
+test('preprocesses headers', () => {
+  let details = {
+    responseHeaders: [
+      {
+        name: 'server',
+        value: 'cloudflare'
+      }
+    ]
+  };
+
+  let r = new Request(details);
+
+  expect(r.headers).toMatchObject({
+    SERVER: 'cloudflare'
+  });
+});
+
+describe('Railgun', () => {
+  test('processes Railgun header v1', () => {
+    let details = {
+      responseHeaders: [
+        {
+          name: 'cf-railgun',
+          value: 'normal 96 foobar 9001'
+        }
+      ]
+    };
+
+    let r = new Request(details);
+
+    expect(r.railgunMetaData).toMatchObject({
+      normal: true,
+      id: 'normal',
+      version: '9001',
+      flags: 0x60,
+      messages: [
+        'Dictionary found in memcache',
+        'Restarted broken origin connection'
+      ]
+    });
+  });
+
+  test('processes Railgun header v2', () => {
+    let details = {
+      responseHeaders: [
+        {
+          name: 'cf-railgun',
+          value: 'foobar 46 700 96 9001'
+        }
+      ]
+    };
+
+    let r = new Request(details);
+
+    expect(r.railgunMetaData).toMatchObject({
+      normal: false,
+      id: 'foobar',
+      version: '9001',
+      compression: '54%',
+      time: '700sec',
+      flags: 0x60,
+      messages: [
+        'Dictionary found in memcache',
+        'Restarted broken origin connection'
+      ]
+    });
+  });
+
+  test('determines Railgun', () => {
+    let detailsRG = {
+      responseHeaders: [
+        {
+          name: 'cf-railgun',
+          value: 'normal'
+        }
+      ]
+    };
+    let detailsNRG = {
+      responseHeaders: []
+    };
+
+    let rRG = new Request(detailsRG);
+    let rNRG = new Request(detailsNRG);
+
+    expect(rRG.servedByRailgun()).toBe(true);
+    expect(rNRG.servedByRailgun()).toBe(false);
+  });
+});
+
+test('determines Cloudflare-ness', () => {
```

Comment:
This made me think of how HTTP headers can repeat (`Cookie` being a common one that repeats), hence the headers passed by Chrome is an array.

I found this block of [pre-processing logic](https://github.com/cloudflare/claire/blob/master/source/request.js#L22-L32) where we only preserve the last header. Did I write this? :)

Do we ensure other `Server` headers are removed before requests leave the CF edge? If that's not guaranteed, we might need to tweak some of the pre-processing or CF determination logic.

Reply:
We do ensure only one Server header appears in the response.